packages/salesforce-adapter/src/filters/foreign_key_references.ts

```diff
@@ -0,0 +1,184 @@
+                if (referenceElemId !== undefined) {
+                  return new ReferenceExpression(referenceElemId)
+                }
+                log.debug(`Could not find reference type for ${id(type)} field: ${field.name} foreignKeyDomain: ${foreignKeyDomain}`)
+              }
+              return foreignKeyDomain
+            })
+
+          const failedReferences = field.annotations[FOREIGN_KEY_DOMAIN].filter(
+            (foreignKeyDomain: string | ReferenceExpression) =>
+              !isReferenceExpression(foreignKeyDomain)
+          )
+          if (failedReferences.length > 0) {
+            log.info(`${failedReferences.length}/${
+              field.annotations[FOREIGN_KEY_DOMAIN].length
+            } foreignKeyDomains for ${id(type)} field ${field.name} failed to resolve: ${
+              failedReferences.join(',')
+            }`)
+          }
+        })
+    })
+}
+
+/**
+ * Generate a mapping from object types to the contained fields that can be resolved
+ * to references, and the choices of referenced metadata types.
+ *
+ * @param elements  The fetched elements
+ */
+const getTypeToFieldReferences = (elements: Element[]): TypeToFieldReferences => (
+  new Map(
+    elements
+      .filter(isObjectType)
+      .map((type: ObjectType): [string, FieldNameToPotentialReferences] => ([
+        type.elemID.getFullName(),
+        new Map(
+          Object.values(type.fields)
+            .filter((f: Field) => Array.isArray(f.annotations[FOREIGN_KEY_DOMAIN]))
+            .filter((f: Field) => f.annotations[FOREIGN_KEY_DOMAIN].some(isReferenceExpression))
+            .map((f: Field) => [
+              f.name,
+              f.annotations[FOREIGN_KEY_DOMAIN].filter(isReferenceExpression).map(
+                (reference: ReferenceExpression): ElemID => reference.elemId
+              ),
+            ])
+        ),
+      ]))
+      .filter((entry: [string, FieldNameToPotentialReferences]) => entry[1].size > 0)
+  )
+)
+
+/**
+ * Resolve references using the mapping genrated from the Salesforce DescribeValueType API.
+ *
+ * @param instance                The current instance being modified
+ * @param referenceFieldMapping   The known reference fields for the instance's type
+ * @param apiNameToElemIDs        Known element ids, mapped by API name and metadata type
+ */
+const resolveReferences = (
+  instance: InstanceElement,
+  referenceFieldMapping: FieldNameToPotentialReferences,
+  apiNameToElemIDs: ApiNameMapping,
+): void => {
+  const shouldReplace = (field: Field): boolean => (
+    referenceFieldMapping.get(field.name) !== undefined
+  )
+
+  const replacePrimitive = (val: Value, field: Field): Value => {
+    if (val === undefined || !_.isString(val)) {
+      return val
+    }
+
+    const potentialValues = (referenceFieldMapping.get(field.name) || []).map(
+      (referenceTypeID: ElemID) =>
+        apiNameToElemIDs?.[val]?.[referenceTypeID.typeName]
+    ).filter(values.isDefined)
```

Comment:
ummm, so, this will probably negate most of the other comments I left but it seems like there is no reason to create `referenceFieldMapping` in the first place.

seems like this can just be replaced with
```suggestion
    const refTarget = makeArray(field.annotations[FOREIGN_KEY_DOMAIN])
      .filter(isReferenceExpression)
      .map(ref => apiNameToElemIDs[val]?.[ref.elemId.typeName])
      .find(values.isDefined)
    return refTarget !== undefined ? new ReferenceExpression(refTarget) : val
```

Reply:
yeah, after the switch to `transformValues` we really don't need it... oops